skills/types/bool_file.go

```diff
@@ -0,0 +1,140 @@
+// -*- Mode: Go; indent-tabs-mode: t -*-
+
+/*
+ * Copyright (C) 2016 Canonical Ltd
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License version 3 as
+ * published by the Free Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+package types
+
+import (
+	"fmt"
+	"regexp"
+
+	"github.com/ubuntu-core/snappy/skills"
+)
+
+// BoolFileType is the type of all the bool-file skills.
+type BoolFileType struct{}
+
+// String() returns the same value as Name().
+func (t *BoolFileType) String() string {
+	return t.Name()
+}
+
+// Name returns the name of the bool-file type (always "bool-file").
+func (t *BoolFileType) Name() string {
+	return "bool-file"
+}
+
+var boolFileGPIOValuePattern = regexp.MustCompile(
+	"^/sys/class/gpio/gpio[0-9]+/value$")
+var boolFileAllowedPathPatterns = []*regexp.Regexp{
+	// The brightness of standard LED class device
+	regexp.MustCompile("^/sys/class/leds/[^/]+/brightness$"),
+	// The value of standard exported GPIO
+	boolFileGPIOValuePattern,
+}
+
+// SanitizeSkill checks and possibly modifies a skill.
+// Valid "bool-file" skills must contain the attribute "path".
+func (t *BoolFileType) SanitizeSkill(skill *skills.Skill) error {
+	if t.Name() != skill.Type {
+		panic(fmt.Sprintf("skill is not of type %q", t))
+	}
+	path, ok := skill.Attrs["path"].(string)
+	if !ok || path == "" {
+		return fmt.Errorf("bool-file must contain the path attribute")
+	}
+	for _, pattern := range boolFileAllowedPathPatterns {
+		if pattern.MatchString(path) {
+			return nil
+		}
+	}
+	return fmt.Errorf("bool-file can only point at LED brightness or GPIO value")
+}
+
+// SanitizeSlot checks and possibly modifies a skill slot.
+func (t *BoolFileType) SanitizeSlot(skill *skills.Slot) error {
+	if t.Name() != skill.Type {
+		panic(fmt.Sprintf("skill slot is not of type %q", t))
+	}
+	// NOTE: currently we don't check anything on the slot side.
+	return nil
+}
+
+// SkillSecuritySnippet returns the configuration snippet required to provide a bool-file skill.
+// Producers gain control over exporting, importing GPIOs as well as
+// controlling the direction of particular pins.
+func (t *BoolFileType) SkillSecuritySnippet(skill *skills.Skill, securitySystem skills.SecuritySystem) ([]byte, error) {
+	switch securitySystem {
+	case skills.SecurityApparmor:
+		// To provide GPIOs we need extra permissions to export/unexport and to
+		// set the direction of each pin.
+		if t.isGPIO(skill) {
+			snippet := []byte(`
```

Comment:
Maybe have this `[]byte` defined at the top of the function instead of inside the `switch` statement?